src/thunderbots/software/grsim_communication/motion_controller.cpp

```diff
@@ -39,19 +46,47 @@ MotionController::Velocity MotionController::bangBangVelocityController(
     }
     else
     {
-        robot_velocities.linear_velocity = MotionController::determineLinearVelocity(
-            robot, dest, desired_final_speed, delta_time, max_speed_meters_per_second,
-            max_angular_speed_radians_per_second);
-        robot_velocities.angular_velocity = MotionController::determineAngularVelocity(
-            robot, desired_final_orientation, delta_time,
-            max_angular_speed_radians_per_second,
-            max_angular_acceleration_meters_per_second_squared);
+        std::visit(
+            overload{
+                [&robot_velocities, robot, delta_time,
+                 this](MotionController::PositionCommand& command) {
+                    robot_velocities.linear_velocity =
+                        MotionController::determineLinearVelocityFromPosition(
+                            robot, command.global_destination,
+                            command.final_speed_at_destination, delta_time,
+                            this->max_speed_meters_per_second,
+                            this->max_angular_speed_radians_per_second);
+                    robot_velocities.angular_velocity =
+                        MotionController::determineAngularVelocityFromPosition(
+                            robot, command.final_orientation, delta_time,
+                            this->max_angular_speed_radians_per_second,
+                            this->max_angular_acceleration_meters_per_second_squared);
+                },
+
+                [&robot_velocities, robot, delta_time,
+                 this](MotionController::VelocityCommand& command) {
+                    if (command.linear_velocity.len() > this->max_speed_meters_per_second)
+                    {
+                        throw std::invalid_argument(
+                            "Desired speed is above the allowed max speed");
+                    }
+                    robot_velocities.linear_velocity =
+                        MotionController::determineLinearVelocityFromVelocity(
+                            robot, command.linear_velocity,
+                            this->max_acceleration_meters_per_second_squared, delta_time);
+                    robot_velocities.angular_velocity =
+                        MotionController::determineAngularVelocityFromVelocity(
+                            robot, command.angular_velocity,
+                            this->max_angular_acceleration_meters_per_second_squared,
+                            delta_time);
+                }},
+            motion_command);
```

Comment:
Might also want to add a case like:
```
[](auto arg){
    throw std::invalid_argument("Unrecognized type passed via std::variant");
}
```
so we throw an exception if we somehow pass an argument type to the `MotionController` that we don't recognize (I think `auto` will match anything). Not sure what would happen if we don't handle a type here, but I'm imagining it's gonna be some sort of nasty C++ error.